api/runtime/hooks/v1alpha1/inplaceupdate_types.go

```diff
@@ -0,0 +1,136 @@
+	// Examples: ["spec.version", "spec.infrastructureRef.spec.memoryMiB"]
+	// +required
+	Changes []string `json:"changes"`
+}
+
+var _ ResponseObject = &CanUpdateMachineResponse{}
+
+// CanUpdateMachineResponse is the response of the CanUpdateMachine hook.
+// +kubebuilder:object:root=true
+type CanUpdateMachineResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains Status and Message fields common to all response types.
+	CommonResponse `json:",inline"`
+
+	// acceptedChanges is the subset of requested changes that this extension can handle.
+	// If empty, the extension cannot handle any of the requested changes.
+	// +optional
+	AcceptedChanges []string `json:"acceptedChanges,omitempty"`
+}
+
+// CanUpdateMachine is the hook that will be called to determine if an extension
+// can handle specific machine changes for in-place updates.
+func CanUpdateMachine(*CanUpdateMachineRequest, *CanUpdateMachineResponse) {}
+
+// UpdateMachineRequest is the request of the UpdateMachine hook.
+// This hook is called to perform the actual in-place update on a machine.
+// +kubebuilder:object:root=true
+type UpdateMachineRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonRequest contains fields common to all request types.
+	CommonRequest `json:",inline"`
+
+	// machineRef is a reference to the machine object the in-place update hook corresponds to.
+	// Updaters should fetch the latest machine state using this reference.
+	// +required
+	MachineRef corev1.ObjectReference `json:"machineRef"`
+}
+
+var _ RetryResponseObject = &UpdateMachineResponse{}
+
+// UpdateMachineStatus represents the status of an in-place update operation.
+// +kubebuilder:validation:Enum=InProgress;Done;Failed
+type UpdateMachineStatus string
+
+const (
+	// UpdateMachineStatusInProgress indicates the update is still in progress.
+	UpdateMachineStatusInProgress UpdateMachineStatus = "InProgress"
+	// UpdateMachineStatusDone indicates the update has completed successfully.
+	UpdateMachineStatusDone UpdateMachineStatus = "Done"
+	// UpdateMachineStatusFailed indicates the update has failed.
+	UpdateMachineStatusFailed UpdateMachineStatus = "Failed"
+)
+
+// UpdateMachineResponse is the response of the UpdateMachine hook.
+// +kubebuilder:object:root=true
+type UpdateMachineResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonRetryResponse contains Status, Message and RetryAfterSeconds fields.
+	CommonRetryResponse `json:",inline"`
+
+	// updateStatus indicates the current status of the update operation.
+	// +required
+	UpdateStatus UpdateMachineStatus `json:"updateStatus"`
+}
+
+// UpdateMachine is the hook that will be called to perform in-place updates on a machine.
+// This hook should be idempotent and can be called multiple times for the same machine
+// until it reports Done or Failed status.
+func UpdateMachine(*UpdateMachineRequest, *UpdateMachineResponse) {}
+
+func init() {
```

Comment:
We probably need to start working to a book page with this documentation, but considering there are still details to be figured out, I'm also ok to defer